test/thunkify.js

```diff
@@ -0,0 +1,28 @@
+var R = require('..');
+var eq = require('./shared/eq');
+
+describe('thunkify', function() {
+  it('returns a function with the same arity as the given function', function() {
+    var input = function input(a0, a1) { };
+    var f = R.thunkify(input);
+    eq(typeof f, 'function');
+    eq(f.length, input.length);
+  });
+
+  it('returns a function that expects arguments and returns a new invoker function', function() {
+    var input = function input(a0, a1) { };
+    var f = R.thunkify(input);
+    eq(typeof f(42, 'xyz'), 'function');
+  });
+
+  it('calls the original function with the provided arguments', function() {
+    var f = R.thunkify(R.add(2));
+    eq(f(42)(), 44);
+  });
+
+  it('should support sending partial arguments', function() {
```

Comment:
I am not sure whether this would be the desired behavior, and it should be discussed. One could argue the function returned by thunkify should be curried, and not return a thunk until all the arguments have been accepted...

Reply:
Not objecting to that, really. It could very well be appropriate, conceptually fitting way to do it.

I'd only like to point out that my original intention was to make the _thunkified_ function behave as close as the one you provide as possible. In that sense, these would be equivalent:

```js
const fn = (a, b) => { ... };

fn(a) === thunkify(fn)(a)();
```

Whereas if I understood you correctly and we decide on currying the result, that would no longer hold true.
 
The thunk only serves as a "delayer" of the invocation and nothing else: a placeholder for a future call.

Reply:
When did I supply a version?  I don't recall doing so.

Ramda generally pays strict attention to the arity of the functions it deals with.  We are loathe to apply only `(a)` to a function expecting `(a, b)`.

Maybe our best bet would be to have a function that supplies a thunk, and not one that creates thunk-generating functions, something like

```
const myThunk = thunkify(expensiveCalc, [a, b])
// ... later
const result = myThunk() //=> expensiveCalc(a, b)
```

Then if you supply fewer arguments than the function has parameters, it's clearly on you.

The more I think about it, the better I like this alternative.

Reply:
> When did I supply a version? I don't recall doing so.
I was referring to your `curry + 1` suggestion. _That_ version.

>  We are loathe to apply only (a) to a function expecting (a, b).
Fair enough. I can also see the benefits of the reasoning. Let's stick to that perspective.

> Maybe our best bet would be to have a function that supplies a thunk, and not one that creates thunk-generating functions
What I don't find attractive about going in that direction is the fact that you lose the ability to simply provide parameters at a later stage if you don't know them at thunk-creation time.

```js
// What if [a, b] are not known here?
const myThunk = thunkify(expensiveCalc, [a, b]);

// I'd like to see something closer to
const myThunk = thunkify(expensiveCalc)([a, b]);
```

Reply:
@nfantone because all Ramda functions are curried by default, you'd be able to do either one of the examples you gave.

Reply:
@Bradcomp Certainly. But that's why I wrote "closer to". The semantics of the first example are, to me at least, clear enough; while on the latter, they seem kinda obscure (is `expensiveCalc` expecting a single array or two arguments `a`, `b`?).

Just a thought 👇 

```js
// IMHO, this:
R.thunkify(R.identity)(42)();

// ...looks much cleaner and predictable than:
R.thunkify(R.identity)([42])(); // <- Why is `R.identity` not returning what I'm passing?
```

Reply:
> ```js
> // What if [a, b] are not known here?
> const myThunk = thunkify(expensiveCalc, [a, b]);
> ```   

So then you're introducing three separate moments in time: one when you know the baseline function you want to thunkify, a second when you know the arguments you want to provide to it but have lost reference to the function itself, and a third when you want to actually apply the function to the arguments and might no longer know either of them.  Is that actually a realistic scenario?

I ask because `thunkify`, even with two distinct moments in time is already an oddity in Ramda, which tries to stick with a functional paradigm.  Functional programming is all about evaluating expressions, and generally you can leave it to the underlying engine to decide when to evaluate individual ones; you don't control sequencing directly.  The existence of [`Monads`][mo]s is in good part about allowing functional programming to actually control some of this sequencing.  In any case, having three separate moments involved seems overkill to me.  But I could be convinced by a good example.

My suggestion of `curry + 1` was off-the-cuff, and while I can see some sense in it, I'm not sure I would really want that API.  I think this demonstrates some issues with it:


```js
const foo1 = curry((a, b) => a + 2 * b)
foo1.length //=> 2
const bar1 = foo1(12)
bar1.length //=> 1
const baz1 = bar1(15)
baz1 //=> 42

const foo2 = thunkify((a, b) => a + 2 * b)
foo2.length //=> ???  2?  3?
const bar2 = foo2(12)
bar2.length //=> ???  1?  2?
const baz2 = bar2(15)
baz2.length //=> ???  0?  1?
baz2() //=> foo2(10, 15) => 42
```

  [mo]: https://github.com/fantasyland/fantasy-land#monad

Reply:
@CrossEye I understand your concern about multiple points in time - but believe  you've stretched the interpretation.

> So then you're introducing three separate moments in time

Actually, only two: _thunk creation time_ and _thunk invocation_. I'm struggling with finding a real world example that fits your three moments.

> Is that actually a realistic scenario?

Not that I can think of, no. But I'm sure there's _bound to be one_ if we think harder.

> My suggestion of curry + 1 was off-the-cuff, and while I can see some sense in it, I'm not sure I would really want that API. 

You raised good points with your arity oddities. The more I think about it, the more I keep coming back to `thunkify(expensiveCalc, [a, b]);` as the better approach.

Reply:
@CrossEye @Bradcomp I'm ready to push this forward and (re-)implement a solution. Are we all inclining to `thunkify(fn, [a, b])` sig here? Any objections/comments?

Reply:
I like this one, and I don't see an issue with the length like @CrossEye does:

```
const foo2 = thunkify((a, b) => a + 2 * b)
foo2.length //=> ???  2?  3?
const bar2 = foo2(12)
bar2.length //=> ???  1?  2?
const baz2 = bar2(15)
baz2.length //=> ???  0?  1?
baz2() //=> foo2(10, 15) => 42
```

The left numbers are correct. A nullary function is a perfectly valid return type, and has a length of zero. Even when taking `curry` into account, the length of a function says nothing of the return type.

The best case scenario in my mind is that `thunkify` does two things to the input function - curries the function, and changes the return type from `b` to `() -> b`.

Reply:
OK, I think I'm convinced.  Describing the thunk as the output and not as one more curried step made me see the light.

At this point, the only thing I see left is that it makes sense to curry `createThunk` instead of simply fixing its arity, that is:

```js
var thunkify = _curry1(function thunkify(fn) {
  return curryN(fn.length, function createThunk() {
    const fnArgs = arguments;
    return function invokeThunk() {
      return fn.apply(this, fnArgs);
    };
  });
});
```

This allows either of these:

```js
thunkify(R.add)(42)(2)() //=> 44
thunkify(R.add)(42, 3)() //=> 45
```

which seems more in line with Ramda.

Reply:
Done.

Reply:
Done.